core/extensions.ts

```diff
@@ -27,7 +27,10 @@ export const TEST_ONLY = {allExtensions};
  * @throws {Error} if the extension name is empty, the extension is already
  *     registered, or extensionFn is not a function.
  */
-export function register(name: string, initFn: Function) {
+export function register<T extends Block>(
+  name: string,
+  initFn: (this: T) => void,
+) {
```

Comment:
This appears to violate [the Golden Rule of Generics](https://effectivetypescript.com/2020/08/12/generics-golden-rule/).  Since T is used only once I believe it can be simplified omit T entirely:

```suggestion
export function register(
  name: string,
  initFn: (this: Block) => void,
) {
```

Reply:
That doesn't work because of the types in the blocks files. They expect subtypes of Block that have their additional methods. The generic is necessary.

Reply:
Oh, hmm.  Correct as usual, but I notice that I am now confused.  I wondered at first if this T was sneakily being a synonym for `any`, but it isn't—yet curiously enough it does not actually enforce the `extends Block`.  For example, this code compiles fine:

```JS
class Foo {};
function init(this: Foo) {}
register('Foo', init);
```

but it enforces _something_, as  this code:

```JS
class Foo {foo = ''}
function init(this: Foo) {this.foo = 'foo';}
register('Foo', init);
```
generates the following error message:

```JS
error TS2345: Argument of type '(this: Foo) => void' is not assignable to parameter of type '(this: Block) => void'.
  The 'this' types of each signature are incompatible.
    Property 'foo' is missing in type 'Block' but required in type 'Foo'.

    register('Foo', init);
                    ~~~~

      class Foo {foo = ''}
                 ~~~
     'foo' is declared here.
```
which is correct— but is barely any less true when one adds `extends Block` to the definition of `class Foo`.

(As an aside, I also discovered that apparently methods do _not_ have an implicit `this` type, much to my surprise, so this might not have come to light if we defined extensions via mixin classes with an init method, rather than plain functions with an explicit `this:` type.)